pandas/_libs/tslibs/nattype.pyi

```diff
@@ -15,7 +18,12 @@ nat_strings: set[str]
 
 def is_null_datetimelike(val: object, inat_is_null: bool = ...) -> bool: ...
 
-class NaTType(datetime):
```

Comment:
`NaTType` inherits from `datetime`. Removing the inheritance because it removes a few mypy ignores doesn't sound like a good idea to me. If there is a reason why `NaTType` shouldn't inherit from `datetime`, the implantation should change.

Reply:
I don't think `NaTType` should inherit from `datetime`, but maybe that's a separate PR?  My reasoning why it should get removed is as follows.  With the current code, the following happens:
```python
>>> type(pd.Timestamp(np.nan))
<class 'pandas._libs.tslibs.nattype.NaTType'>
>>> type(pd.Timedelta(np.nan))
<class 'pandas._libs.tslibs.nattype.NaTType'>
True
```
So here we are saying that the `Timedelta` constructor can produce an object of type `NaTType` which is also a `datetime`, but `datetime` and `timedelta` are different in the python standard library.

From a static typing perspective, it is messing up a number of different things, and I don't think we get any benefit from that inheritance with respect to typing.

Furthermore, as best as I can tell, the `NaTType` implementation just overrides the entire `datetime` API, so we are not getting any implementation benefits from explicitly stating the inheritance.  But I could be missing something here.

Reply:
> NaTType inherits from datetime. Removing the inheritance because it removes a few mypy ignores doesn't sound like a good idea to me. If there is a reason why NaTType shouldn't inherit from datetime, the implantation should change.

While in the `nattype.pyx` part of the code, we indicate this inheritance, from a typing perspective, it's not really the case because there are methods in `datetime` that are not implemented in `NaTType`.  We even have a test to check that: `pandas/tests/scalar/test_nat.py:test_missing_public_nat_methods()` 

With static typing, the assumption is that a class that is a subclass of the parent has all the methods of the parent.  Our implementation of `NaTType` doesn't do that, so we need to tell the type system that `NaTType` is not a subclass of `datetime`

Reply:
Wouldn't changing the annotation without changing the implementation mess up static inference on e.g.:

```
if isinstance(obj, datetime):
    return 0
else:
    assert False
```

Reply:
> methods in `datetime` that are not implemented in `NaTType`. We even have a test to check that: `pandas/tests/scalar/test_nat.py:test_missing_public_nat_methods()`

I'm really looking forward to #24983 (which might take some time).

If Timestamp/Timedelta/NaTType were the last pieces to get pandas to be a py.typed library, I would fully agree to purposefully make the two large changes proposed in this PR (not inherit from `datetime` and simplifying `__new__`)  - the benefits for users would (from my perspective) outweigh the discrepancy between implementation and typing. With the current state of Pandas it is more difficult to judge: I'm not against this PR but also not in favor of it.

Reply:
> Wouldn't changing the annotation without changing the implementation mess up static inference on e.g.:
> 
> ```
> if isinstance(obj, datetime):
>     return 0
> else:
>     assert False
> ```

It really depends on how you are handling `pd.NaT` in your user code, whether you are talking about `Timestamp` versus `Timedelta`, etc.

As pandas users, my staff and I live in the `Timestamp`/`Timedelta` world.  We use `pd.to_datetime()` to convert strings to dates and times and get `Timestamp` objects as a result.   The relationships of those types to `datetime.datetime` and `datetime.timedelta` doesn't matter.  So I wouldn't write code like that example.  Besides, we have this kind of inconsistency to worry about, independent of typing:
```python
>>> isinstance(pd.Timestamp("NAN"), datetime.datetime)
True
>>> isinstance(pd.Timestamp("NAN"), pd.Timestamp)
False
```
Seems kind of odd that you ask for a `Timestamp`, which is a subclass of `datetime`, and the result is `True` to ask if it is a `datetime`, but is `False` when asking if it is a `Timestamp`

So the point is that the implementation has issues - that's maybe what #24983 would take care of, but until then we should type it in a way that makes sense from a user perspective.

For what it's worth, we don't document that `Timestamp` is a subclass of `datetime.datetime` nor do we even have a documentation page about `pd.NaT` at all that indicates anything about its type.

Reply:
> For what it's worth, we don't document that Timestamp is a subclass of datetime.datetime

There's a note about it here in the API reference but probably easy to miss: https://pandas.pydata.org/pandas-docs/stable/reference/arrays.html#datetime-data